Fixing Java field access

oxisto opened the pull request:
This PR consists of a major overhaul of Java reference and member expression parsing. It removes a LOT of old weird legacy code that was full of weird things.

It also introduces a new Java-specific pass, that takes care of deciding when a member access is actually static or not. This might be something that we could decide to adapt for other/all languages (maybe also tied to the discussion in #1863).

oxisto commented:
Blocked by #1586

KuechA commented on cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/ExpressionHandler.kt:
```diff
@@ -502,111 +360,22 @@ class ExpressionHandler(lang: JavaLanguageFrontend) :
-        //    }
-        val name = this.parseName(nameExpr.nameAsString)
-        return try {
+        // Try to resolve it. We will remove this in a future where we do not really in the
```
Is the `TODO` of this method resolved? If so, please delete it.

Also, please document the method according to the new standards: https://github.com/Fraunhofer-AISEC/cpg/pull/1683#discussion_r1752563555

This sentence is broken

KuechA commented on cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/ExpressionHandler.kt:
```diff
@@ -244,171 +242,31 @@ class ExpressionHandler(lang: JavaLanguageFrontend) :
-                    ?: newProblemExpression("Could not parse base")
-        }
-        var fieldType: Type?
+
```
Please document the method according to the new standards: https://github.com/Fraunhofer-AISEC/cpg/pull/1683#discussion_r1752563555

KuechA commented on cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/JavaExtraPass.kt:
```diff
@@ -0,0 +1,83 @@
+
+@DependsOn(TypeResolver::class)
+@ExecuteBefore(SymbolResolver::class)
+class JavaExtraPass(ctx: TranslationContext) : TranslationUnitPass(ctx) {
```
Documentation missing for this whole file

KuechA reviewed (commented):
I fail to understand most of the changes since the old code is hard to understand and the new code requires in-depth knowledge and understanding of the java parser and frontend which I do not have. However, I have some nits which should already be resolved and may support subsequent more detailed reviews.

oxisto commented on cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:
```diff
@@ -209,7 +209,7 @@ open class SymbolResolver(ctx: TranslationContext) : ComponentPass(ctx) {
         // ideal since we are running into an issue with function pointers here (see workaround
         // below).
-        var wouldResolveTo = ref.candidates.singleOrNull()
+        var wouldResolveTo = ref.candidates.firstOrNull()
```
It was semi-intended. I am ok with the old behavior, but we should at least
1) have a warning if we do not resolve it
2) forward the decision to the `Language`. In Java it is possible to have an ambiguity here if you statically import a field and a function and both can have the same name and end up in candidates. I need to deal with that here

oxisto commented on cpg-language-java/src/test/kotlin/de/fraunhofer/aisec/cpg/enhancements/calls/SuperCallTest.kt:
```diff
@@ -89,13 +89,16 @@ internal class SuperCallTest : BaseTest() {
+        // See https://github.com/Fraunhofer-AISEC/cpg/issues/1863.
+        // We only have a regular call expression here, if we do not resolve anything else in the
+        // frontend
+        /*assertTrue(fieldRef.base is Reference)
```
We expect the lines below. Fixed.

oxisto commented:
> I fail to understand most of the changes since the old code is hard to understand and the new code requires in-depth knowledge and understanding of the java parser and frontend which I do not have. However, I have some nits which should already be resolved and may support subsequent more detailed reviews.

Please re-check, I addressed the issues.
